### PR TITLE
Cloud Native - Chapter 5 - Updates to gRPC

### LINKED ISSUE #17532: Title: gRPC instead of REST?
The title says REST and gRPC but REST is only mentioned in the intro and the rest (pardon the pun) is gRPC. I think the title could better reflect this.

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: dc86015f-c6c7-5fdc-ae0f-c69353afcb34
* Version Independent ID: 491d7006-ae8d-fb74-8b29-7e4b5031fcae
* Content: [REST and gRPC](https://docs.microsoft.com/en-us/dotnet/architecture/cloud-native/rest-grpc)
* Content Source: [docs/architecture/cloud-native/rest-grpc.md](https://github.com/dotnet/docs/blob/master/docs/architecture/cloud-native/rest-grpc.md)
* Product: **dotnet-architecture**
* Technology: **cloud-native**
* GitHub Login: @robvet
* Microsoft Alias: **nanil**

### PR DESCRIPTION
**Pull request by robvet**
## Summary

Update to gRPC section in Chapter 5 in response to a user request.
Also closes #17532 #14720

### CODE REVIEW

**Comment by nishanil** [comment]
Referencing issue: #14720

**Comment by nishanil** [comment]
@robvet The ask is more around having concrete examples of gRPC usage in microservices-to-microservices communication. Probably, picking up BFF/aggregators as an example scenario and expanding on them will be ideal. The diagram also should be updated accordingly. 
eShop sample uses them in BFF. Check these links:
* https://github.com/dotnet-architecture/eShopOnContainers/wiki/BFF-implementation
* https://github.com/dotnet-architecture/eShopOnContainers/wiki/API-gateways 
* https://github.com/dotnet-architecture/eShopOnContainers/wiki/gRPC

**Comment by mairaw** [comment]
There are merge conflicts here

**Comment by nishanil** [comment]
@robvet Are you planning to cover some examples of gRPC usage as requested in the issue referenced?

**Comment by robvet** [comment]
Hey Nish.. on break from customer. Yes, would like to show some code excerpts from eshop.

As we move forward with book, hoping to show more code examples.

Thoughts?

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________

File: docs/architecture/cloud-native/rest-grpc.md [hunk 1]
```diff
@@ -1,69 +1,347 @@
...
 
 ## Protocol Buffers
 
-gRPC embraces another open-source technology called [Protocol Buffers](https://developers.google.com/protocol-buffers/docs/overview) or Protobuf messages to send and receive data. Similar to a [WCF Data Contract](https://docs.microsoft.com/dotnet/framework/wcf/feature-details/using-data-contracts), Protobuf serializes structured data for systems to read and write. It reduces the overhead that human-readable formats like XML or JSON incur.
+gRPC embraces an open-source technology called [Protocol Buffers](https://developers.google.com/protocol-buffers/docs/overview). They provide a highly efficient and platform-neutral serialization format for serializing structured messages that services send to each other. Using a cross-platform Interface Definition Language (IDL), developers define a service contract for each microservice. The contract, implemented as a text-based '.proto' file, describes the methods, inputs, and outputs for each service. The same contract file can be used for gRPC clients and services built on different development platforms.
```
Comment (nishanil) [inline docs/architecture/cloud-native/rest-grpc.md]:
```suggestion
gRPC embraces an open-source technology called [Protocol Buffers](https://developers.google.com/protocol-buffers/docs/overview). They provide a highly efficient and platform-neutral serialization format for serializing structured messages that services send to each other. Using a cross-platform Interface Definition Language (IDL), developers define a service contract for each microservice. The contract, implemented as a text-based `.proto` file describes the methods, inputs, and outputs for each service. The same contract file can be used for gRPC clients and services built on different development platforms.
```

File: docs/architecture/cloud-native/rest-grpc.md [hunk 2]
```diff
@@ -1,69 +1,347 @@
...
 
-Many object serialization techniques reflect across the structure of data objects at run-time. Protobuf requires you to define the structure up front with a platform-agnostic language (Protocol Buffer Language). Each definition is stored in a ".proto" file. Then using Protobuf compiler, "Proton," you generate client and server code for any of the supported platforms. The generated code is optimized for fast serialization/deserialization of data. At runtime, each message is wrapped in the strongly-typed service contract and serialized in a standard Protobuf representation.
+Using the proto file," the Protobuf compiler, "Protoc," generates both client and service code for your target platform. The code includes the following components:
```
Comment (nishanil) [inline docs/architecture/cloud-native/rest-grpc.md]:
```suggestion
Using the proto file, the Protobuf compiler, `protoc`, generates both client and service code for your target platform. The code includes the following components:
```

File: docs/architecture/cloud-native/rest-grpc.md [hunk 3]
```diff
@@ -1,69 +1,347 @@
...
+
+At runtime, each message is serialized as a standard Protobuf representation and exchanged between the client and remote service. Unlike JSON or XML, Protobuf messages are serialized as compiled binary bytes.
+
+The book, [gRPC for WCF Developers](https://docs.microsoft.com/dotnet/architecture/grpc-for-wcf-developers/), available from the Microsoft Architecture Site, provides in-depth coverage of gRPC and Protocol Buffers.
 
 ## gRPC support in .NET
 
-The Microsoft .NET Core framework 3.0 includes tooling and native support for gRPC. Figure 4-20 shows the Visual Studio 2019 template that scaffolds a gRPC skeleton project for a gRPC service. Note how .NET Core supports the Windows, Linux, and macOS platforms.
+Microsoft .NET Core 3.0 includes tooling and native support for gRPC. gRPC has been seamlessly integrated into the 3.0 framework with support for endpoint routing, built-in IoC, and logging. The open-source Kestrel web server supports HTTP/2 connections. Figure 4-20 shows a Visual Studio 2019 template that scaffolds a skeleton project for a gRPC service. Note how .NET Core fully supports the Windows, Linux, and macOS platforms.
```
Comment (nishanil) [inline docs/architecture/cloud-native/rest-grpc.md]: Rephrase this entire para. The use of the word `framework` is confusing in the context of .NET Core. Use `.NET Core 3.0 SDK`
Also, 
- mention gRPC is supported in .NET Core 3.0 and **later**. 
- Visual Studio 2019 version 16.3 or later, with the ASP.NET and web development workload installed.
- support in dotnet CLI 

The sentence is confusing:

> Note how .NET Core fully supports the Windows, Linux, and macOS platforms.

File: docs/architecture/cloud-native/rest-grpc.md [hunk 4]
```diff
@@ -1,69 +1,347 @@
...
+
+Favor gRPC for the following scenarios:
+
+- Synchronous backend microservice-to-microservice communication where an immediate response is required to continue processing.
+- Polyglot environments that need to support mixed programming platforms.
+- Low latency and high throughput communication where performance is critical.
+- Point-to-point real-time communication - gRPC can push messages in real time without polling and has excellent support for bi-directional streaming.
+- Network constrained environments – binary gRPC messages are always smaller than an equivalent text-based JSON message.
+
+At the time of this writing, gRPC is primarily used in backend services. Most modern browsers can't provide the level of HTTP/2 control required to support a front-end gRPC client.
```
Comment (nishanil) [inline docs/architecture/cloud-native/rest-grpc.md]: You can also mention about the experimental support for gRPC-Web: https://devblogs.microsoft.com/aspnet/grpc-web-experiment/

File: docs/architecture/cloud-native/rest-grpc.md [hunk 5]
```diff
@@ -1,69 +1,347 @@
...
+
+## gRPC implementation
+
+The microservice reference architecture for [eShop on Containers](https://github.com/dotnet-architecture/eShopOnContainers) demonstrates how to implement gRPC into microservices on the .NET Core platform. Figure 4-22 shows the implementation.
+
+![gRPC in eShop on Containers](./media/grpc-implementation.png)
```
Comment (nishanil) [inline docs/architecture/cloud-native/rest-grpc.md]: Use a better diagram here. The important aspect here is to show how aggregators use gRPC.

Comment (nishanil) [reply docs/architecture/cloud-native/rest-grpc.md]: You can briefly mention the service mesh implementation too. Those are practical scenarios.

File: docs/architecture/cloud-native/rest-grpc.md [hunk 6]
```diff
@@ -1,69 +1,347 @@
...
+
+**Figure 4-22**. gRPC in eShop on Containers
+
+In the previous figure, note how eShop exposes multiple API gateways (on the left) embracing the [Backend for Frontends pattern](https://docs.microsoft.com/azure/architecture/patterns/backends-for-frontends). The *Web-Marketing gateway* (on the bottom) implements RESTful HTTP services for simple CRUD operations. It also includes an Aggregator microservice that implements a gRPC client. The gRPC client makes synchronous gRPC-based calls (in red) to backend gRPC services for more complex operations. Keep in mind that gRPC communication requires both client and server components.
+
+### gRPC client implementation
```
Comment (nishanil) [inline docs/architecture/cloud-native/rest-grpc.md]: IMHO, Client and server implementation details are unnecessary here. The Get-Started guides that you have linked will cover those. The book should be focusing on gRPC scenarios only.

File: docs/architecture/cloud-native/rest-grpc.md [hunk 7]
```diff
@@ -1,69 +1,347 @@
...
+    }
+    ...
+}
+```
+
+**Figure 4-26** Custom gRPC client code
+
+Note in the previous figure how *GetById* method in the *BasketService* class creates an anonymous method. The method is passed into the generic *CallService* method in the *GrpcCallerService*. Importantly, you can see how *CallService* consumes the system-generated *Basket.BasketClient* and *gRPChannel* gRPC plumbing code to invoke remote calls.
+
+### gRPC server implementation
```
Comment (nishanil) [inline docs/architecture/cloud-native/rest-grpc.md]: As mentioned earlier, Server & Client implementation details can be avoided in this book.

File: docs/architecture/cloud-native/rest-grpc.md [hunk 8]
```diff
@@ -1,69 +1,347 @@
...
+```
 
-**Figure 4-22**. gRPC usage patterns
+Note in the previous figure how two endpoints are opened. One listens for HTTP traffic for the RESTful endpoints and the other for gRPC calls. Importantly, the gRPC endpoint specifies the HTTP/2 protocol which is required for gRPC communication.
 
-Note in the previous figure how front-end traffic is invoked with HTTP while back-end microservice to microservice uses gRPC.
+## Looking ahead
 
-Looking ahead, gRPC could play a major role in dethroning the dominance of REST for cloud-native systems. The performance benefits and ease of development are too good to pass up. However, make no mistake, REST will still be around for a long time. It still excels for publicly exposed APIs and for backward compatibility reasons.
+Looking ahead, gRPC could well play a major role in dethroning the dominance of REST for cloud-native systems. The performance benefits and ease of development are compelling. However, REST will still most likely be around for a long time. It excels for publicly exposed APIs and for backward compatibility reasons.
```
Comment (nishanil) [inline docs/architecture/cloud-native/rest-grpc.md]: The use of a word like dethroning is unnecessary. We should be subtle in our prediction statements :)

File: docs/architecture/cloud-native/service-mesh-communication-infrastructure.md [hunk 9]
```diff
@@ -59,14 +59,20 @@ We next move to distributed data in cloud-native systems and the benefits and ch
 
 - [About Ingress in Azure Kubernetes Service (AKS)](https://vincentlauzon.com/2018/10/10/about-ingress-in-azure-kubernetes-service-aks/)
 
-- [Practical gRPC](https://www.worldcat.org/title/practical-grpc/oclc/1042342319)
+- [Practical gRPC](https://www.amazon.com/Practical-gRPC-Joshua-Humphries-ebook/dp/B07DMS44WT/ref=sr_1_1?keywords=Practical+gRPC&qid=1581887357&sr=8-1)
```
Comment (nishanil) [inline docs/architecture/cloud-native/service-mesh-communication-infrastructure.md]: This link has search keywords. Also, I am not sure if we need to link this book? The gRPC official docs are great.

File: docs/architecture/cloud-native/rest-grpc.md (hunk 5 above)
Comment (nishanil) [reply docs/architecture/cloud-native/rest-grpc.md]: Check this : https://github.com/dotnet-architecture/eShopOnContainers/wiki/API-gateways

File: docs/architecture/cloud-native/service-mesh-communication-infrastructure.md (hunk 9 above)
Comment (robvet) [reply docs/architecture/cloud-native/service-mesh-communication-infrastructure.md]: As discussed, removed link to this external book

File: docs/architecture/cloud-native/rest-grpc.md (hunk 8 above)
Comment (robvet) [reply docs/architecture/cloud-native/rest-grpc.md]: Remove "dethroning" and softened prediction.

File: docs/architecture/cloud-native/rest-grpc.md (hunk 3 above)
Comment (robvet) [reply docs/architecture/cloud-native/rest-grpc.md]: Refactored paragraph

File: docs/architecture/cloud-native/CNBook.code-workspace [hunk 10]
```diff
@@ -0,0 +1,8 @@
+{
```
Comment (nishanil) [inline docs/architecture/cloud-native/CNBook.code-workspace]: Why do we need this file?

File: docs/architecture/cloud-native/rest-grpc.md (hunk 7 above)
Comment (robvet) [reply docs/architecture/cloud-native/rest-grpc.md]: Removed and refactored.

File: docs/architecture/cloud-native/rest-grpc.md (hunk 6 above)
Comment (robvet) [reply docs/architecture/cloud-native/rest-grpc.md]: Removed anb refactored.

File: docs/architecture/cloud-native/rest-grpc.md (hunk 7 above)
Comment (robvet) [reply docs/architecture/cloud-native/rest-grpc.md]: Removed

File: docs/architecture/cloud-native/CNBook.code-workspace (hunk 10 above)
Comment (robvet) [reply docs/architecture/cloud-native/CNBook.code-workspace]: Removed

File: docs/architecture/cloud-native/rest-grpc.md (hunk 5 above)
Comment (robvet) [reply docs/architecture/cloud-native/rest-grpc.md]: Refactored.

File: docs/architecture/cloud-native/rest-grpc.md (hunk 4 above)
Comment (robvet) [reply docs/architecture/cloud-native/rest-grpc.md]: Added!

File: docs/architecture/cloud-native/grpc.md [hunk 11]
```diff
@@ -0,0 +1,113 @@
...
+
+## Protocol Buffers
+
+gRPC embraces an open-source technology called [Protocol Buffers](https://developers.google.com/protocol-buffers/docs/overview). They provide a highly efficient and platform-neutral serialization format for serializing structured messages that services send to each other. Using a cross-platform Interface Definition Language (IDL), developers define a service contract for each microservice. The contract, implemented as a text-based '.proto' file, describes the methods, inputs, and outputs for each service. The same contract file can be used for gRPC clients and services built on different development platforms.
+
+Using the proto file, the Protobuf compiler, "Protoc," generates both client and service code for your target platform. The code includes the following components:
```
Comment (nishanil) [inline docs/architecture/cloud-native/grpc.md]:
```suggestion
Using the proto file, the Protobuf compiler, `protoc`, generates both client and service code for your target platform. The code includes the following components:
```

File: docs/architecture/cloud-native/grpc.md [hunk 12]
```diff
@@ -0,0 +1,113 @@
...
+
+gRPC is lightweight and highly performant. It can be up to 8x faster than JSON serialization with messages 60-80% smaller. In Microsoft [Windows Communication Foundation (WCF)](https://docs.microsoft.com/dotnet/framework/wcf/whats-wcf) parlance, gRPC performance exceeds the speed and efficiency of the highly optimized [NetTCP bindings](https://docs.microsoft.com/dotnet/api/system.servicemodel.nettcpbinding?view=netframework-4.8). Unlike NetTCP, which favors the Microsoft stack, gRPC is cross-platform.
+
+## Protocol Buffers
+
+gRPC embraces an open-source technology called [Protocol Buffers](https://developers.google.com/protocol-buffers/docs/overview). They provide a highly efficient and platform-neutral serialization format for serializing structured messages that services send to each other. Using a cross-platform Interface Definition Language (IDL), developers define a service contract for each microservice. The contract, implemented as a text-based '.proto' file, describes the methods, inputs, and outputs for each service. The same contract file can be used for gRPC clients and services built on different development platforms.
```
Comment (nishanil) [inline docs/architecture/cloud-native/grpc.md]:
```suggestion
gRPC embraces an open-source technology called [Protocol Buffers](https://developers.google.com/protocol-buffers/docs/overview). They provide a highly efficient and platform-neutral serialization format for serializing structured messages that services send to each other. Using a cross-platform Interface Definition Language (IDL), developers define a service contract for each microservice. The contract, implemented as a text-based `.proto` file, describes the methods, inputs, and outputs for each service. The same contract file can be used for gRPC clients and services built on different development platforms.
```

File: docs/architecture/cloud-native/grpc.md [hunk 13]
```diff
@@ -0,0 +1,113 @@
...
+
+## gRPC implementation
+
+The microservice reference architecture, [eShop on Containers](https://github.com/dotnet-architecture/eShopOnContainers), from Microsoft, shows how to implement gRPC services in .NET Core applications. Figure 4-22 presents the back-end architecture.
+
+![Backend architecture for eShop on Containers](./media/eshopWithAggregators.png)
+
+**Figure 4-22**. Backend architecture for eShop on Containers
+
+In the previous figure, note how eShop embraces the [Backend for Frontends pattern](https://docs.microsoft.com/azure/architecture/patterns/backends-for-frontends) (BFF) by exposing multiple API gateways. We discussed the BFF pattern in chapter 4. Pay close attention to the Aggregator microservice (in gray) that sits between the Web-Shopping API Gateway and backend Shopping microservices. The Aggregator orchestrates operations that require interaction with multiple backend services. These calls require synchronous communication and an immediate response. In eShop, backend calls from the Aggregator are performed using gRPC as shown in Figure 4-23.
```
Comment (nishanil) [inline docs/architecture/cloud-native/grpc.md]: `The Aggregator orchestrates operations that require interaction with multiple backend services.` 
This staement is confusing. Aggregator does not orchestrate. Consider changing to 
`The Aggregator receives a single request from a client, dispatches them to the various microservices, and then aggregates the results and sends them back to the requesting client.`

File: docs/architecture/cloud-native/grpc.md [hunk 14]
```diff
@@ -0,0 +1,113 @@
...
+
+- Strongly-typed code-generated clients
+- Compact Protobuf messages
+- Server streaming
+
+## gRPC implementation
+
+The microservice reference architecture, [eShop on Containers](https://github.com/dotnet-architecture/eShopOnContainers), from Microsoft, shows how to implement gRPC services in .NET Core applications. Figure 4-22 presents the back-end architecture.
+
+![Backend architecture for eShop on Containers](./media/eshopWithAggregators.png)
```
Comment (nishanil) [inline docs/architecture/cloud-native/grpc.md]: Can you fix the image name here? Standardize to use with hyphens. I think you already have an image with the name `eshop-grpc-wide.png` which can be used here? If yes, drop the `eshopWithAggregators.png` and update the reference.

File: docs/architecture/cloud-native/grpc.md [hunk 15]
```diff
@@ -0,0 +1,113 @@
...
+
+**Figure 4-22**. Backend architecture for eShop on Containers
+
+In the previous figure, note how eShop embraces the [Backend for Frontends pattern](https://docs.microsoft.com/azure/architecture/patterns/backends-for-frontends) (BFF) by exposing multiple API gateways. We discussed the BFF pattern in chapter 4. Pay close attention to the Aggregator microservice (in gray) that sits between the Web-Shopping API Gateway and backend Shopping microservices. The Aggregator orchestrates operations that require interaction with multiple backend services. These calls require synchronous communication and an immediate response. In eShop, backend calls from the Aggregator are performed using gRPC as shown in Figure 4-23.
+
+![gRPC in eShop on Containers](./media/grpc-implementation.png)
```
Comment (nishanil) [inline docs/architecture/cloud-native/grpc.md]: The figure has a typo. Use `HTTP/2` instead of `Https/2`

**Review by nishanil — COMMENTED** [review]
@robvet Found a few things that got missed out.

File: docs/architecture/cloud-native/grpc.md (hunk 14 above)
Comment (robvet) [reply docs/architecture/cloud-native/grpc.md]: Great catch. Done

File: docs/architecture/cloud-native/grpc.md (hunk 15 above)
Comment (robvet) [reply docs/architecture/cloud-native/grpc.md]: Great catch; fixed

File: docs/architecture/cloud-native/grpc.md (hunk 13 above)
Comment (robvet) [reply docs/architecture/cloud-native/grpc.md]: Perfect. Updated

**Review by nishanil — APPROVED** [review]
LGTM! Thanks @robvet

**Pull request merged**